usdigitalresponse/grants-ingest
# Extract XML from grants.gov zip archive downloads

### Ticket #6 

## Description

This PR adds a new `ExtractGrantsGovDBToXML` Lambda function according to #6. Specifically, the Lambda function is triggered when a new Grants.gov zip archive is uploaded to S3 under a key path of `sources/YYYY/mm/dd/grants.gov/archive.zip`; it performs the following actions:
1. Sequentially streams the source (`archive.zip`) file contents from S3 to an in-memory buffer
2. Progressively unzips the buffer contents as it continues to be filled by the downloader
3. Progressively streams the extracted (XML file) contents back to S3

Because zip archives contain "entries" of compressed files and their metadata that are bundled in-place (as opposed to a providing a manifest of contents located at the start of the file), the entries must be progressively discovered in order to understand the full contents of the archive. We expect that Grants.gov zip archives contain exactly one file, which should be named with a `.xml` extension. If, during extraction, additional files are discovered, the invocation fails early to provide an opportunity for a developer to investigate. However, since the upload occurs _while_ the extraction is taking place (in order to control memory usage), we first upload to a temporary key prefix. Once the Lambda handler verifies that the zip archive contains no surprises, the S3 is moved (within the same bucket) to its permanent location next to the source `archive.zip` file (i.e. `sources/YYYY/mm/dd/grants.gov/extract.xml`).

## Testing

1. Follow the normal process for provisioning a LocalStack environment with Terraform.
2. Create a zip file named `archive.zip` that contains a single compressed file entry. The name of the compressed file can be anything as long as it ends in `.xml`. This Lambda function doesn't actually care if the file contains valid XML, so its contents can be anything.
3. Use `awslocal` to upload the zip archive to S3:
    ```cli
    $ awslocal s3 cp --sse AES256 archive.zip \
           s3://grants-ingest-grantssourcedata-000000000000-us-west-2/sources/2023/05/24/grants.gov/archive.zip
    ```
4. Observe that the Lambda function is invoked following the S3 upload. 
5. Once the invocation finishes (after a few moments), check that the extracted XML file has been uploaded to the same base path as the `archive.zip` file:
    ```cli
    $ awslocal s3 ls --recursive \
           s3://grants-ingest-grantssourcedata-000000000000-us-west-2/sources/2023/05/24/grants.gov/
    2023-05-24 16:26:09   1234 sources/2023/05/24/grants.gov/archive.zip
    2023-05-24 16:57:13  98765 sources/2023/05/24/grants.gov/extract.xml
    ```
6. Bonus: Upload another zip file containing multiple compressed file entries. The invocation should fail, and no `extract.xml` file should be present under the path of the source `archive.zip` file (although it may be present under `tmp/` depending on the order of the entries in the archive).

### Automated and Unit Tests

### Manual tests for Reviewer

## Checklist
